src/runtime_src/hip/core/event.cpp

```diff
@@ -4,66 +4,180 @@
 #include "event.h"
 
 namespace xrt::core::hip {
+event::
+event(std::shared_ptr<stream>&& s)
+  : command(std::move(s))
+{
+  ctype = type::event;
+}
+
+void
+event::
+record(std::shared_ptr<stream> s)
+{
+  cstream = std::move(s);
+  auto ev = std::dynamic_pointer_cast<event>(command_cache.get(static_cast<command_handle>(this)));
+  throw_invalid_handle_if(!ev, "event passed is invalid");
+  if (is_recorded()) {
+    // already recorded
+    //cstream->erase_cmd(ev); It is commented will be uncomment after stream PR is pushed
+  }
+  // update recorded commands list
+  //cstream->enqueue_event(std::move(ev));It is commented will be uncomment after stream PR is pushed
+  set_state(state::recorded);
+}
 
 bool
-command::
-submit()
+event::
+is_recorded()
 {
-    return true; //temporary
+  return get_state() >= command::state::recorded;
 }
 
 bool
-command::
-wait()
+event::
+query()
 {
-    return true; //temporary
+  for (auto it = recorded_commands.begin(); it != recorded_commands.end(); it++){
+    state command_state = (*it)->get_state();
+    if (command_state != state::completed){
+      return false;
+    }
+  }
+  return true;
 }
 
 bool
 event::
-submit()
+synchronize()
 {
-    return true; //temporary
+  for (auto it = recorded_commands.begin(); it != recorded_commands.end(); it++){
+    state command_state = (*it)->get_state();
+    if (command_state < state::completed){
+      (*it)->wait();
+      (*it)->set_state(state::completed);
+    }
+  }
+  set_state(state::completed);
+  for (auto it = chain_of_commands.begin(); it != chain_of_commands.end(); it++){
+    (*it)->submit(true);
```

Comment:
We dont require bool for submit. Please remove it. Let stream algorithm take care when to submit.